sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -543,6 +543,13 @@ object ScalaReflection extends ScalaReflection {
             throw new UnsupportedOperationException(s"`$fieldName` is a reserved keyword and " +
               "cannot be used as field name\n" + walkedTypePath)
           }
+          // It is ok to use head here because a variable name needs to be non empty
+          if (!Character.isJavaIdentifierStart(fieldName.head)) {
+            throw new UnsupportedOperationException(s"`$fieldName` is a word that starts with" +
+             " a character not allowed in Java and cannot be used as field name. " +
+              "A workaround to this problem is to convert the case class to Json" + 
+              " and then read the schema as Json\n" + walkedTypePath)
+          }
```

Comment:
```suggestion
          // It is ok to use head here because a variable name needs to be non empty
          if (!Character.isJavaIdentifierStart(fieldName.head)) {
            throw new UnsupportedOperationException(s"`$fieldName` is a word that starts with" +
             " a character not allowed in Java and cannot be used as field name. " +
              "A workaround to this problem is to convert the case class to Json" +
              " and then read the schema as Json\n" + walkedTypePath)
          }
```